### DISCUSSION TITLE
How to add firmware support for connecting to and controlling other Bluetooth Low Energy devices (such as BuWizz 2.0)?

### DISCUSSION THREAD

**Discussion by Mantarri** [pybricks/support#1452] [Q&A]
Hello. I have a generic bluetooth gamepad, BuWizz 2.0, and NXT 2.0 and was hoping to use at least some of these with my Technic hub. I've tried looking to see if there's a more generic bluetooth package from PyBricks so I can try connecting to those devices and communicating them, but so far I've turned up empty-handed.

**Comment by Mantarri** [comment]
So after some discussion on the PyBricks discord channel I found out that none of these are currently possible, but that the NXT brick is being looked into, as well as the possibility of supporting other BLE controllers. My followup questions I'll put here, as well as change the discussion title to reflect them. 

Followup question: what would it take for me to add support on my own for the Powered Up Technic hub to connect to an arbitrary Bluetooth LE device (in this case a BuWizz 2.0) and control it via its Bluetooth API, since the currently exposed BLE API on the Technic hub is only capable of using BLE's broadcasting ability for communication; no pairing-based communication besides the Xbox Wireless controllers with BLE support. 

I assume the code for Xbox controller support would be relevant, so where would I find anything in the codebase related to that (I don't want to accidentally miss anything.)

**Reply by dlech** [reply]
We seem to be missing the relevant tags to get a proper diff range (https://github.com/pybricks/pybricks-micropython/issues/235) but looking at the two pages of commits starting at https://github.com/pybricks/pybricks-micropython/commits/97c33097efa54e9197b726307d4db1cda1e26d4e/ should have all of the relevant code to look at.

**Reply by Mantarri** [reply]
Thanks. Will definitely be taking a look into this. If I manage to get this setup, would that be something that core may be interested, or should I stick as my own personal fork?

**Reply by laurensvalk** [reply]
Your own version would be a good place to start.

The firmware size is quite limited, so instead of adding support for many different devices at the firmware level, I suppose we can look at generalizing some Bluetooth class so you can build your own classes for this sort of thing at the Python level.

I don't know that we'll have all the bells and whistles of a full bluetooth stack, but maybe there's a subset of features we can expose. For the gamepads, maybe HID over BLE is generic enough.

In practice, it's always good to start with something specific like you are doing, which tends to give a nice overview of the features that might be needed.

**Reply by Mantarri** [reply]
Makes sense. What is the firmware size limit, and what is the current size (and is it different for different devices? I only have a Technic hub to work with currently.)

As far as it goes, I was able to get a proof-of-concept for connection and control of my BuWizz 2 hub via ~~PyBluez~~ Bleak, so now I'll be taking a stab at doing it in C on the hub. From there I'd be looking at exposing some functionality to the API. At the very least motor control and BuWizz speed mode (allows running the motors at higher power than the stock Power Functions batteries/controllers) and hopefully allowing access to battery voltage and some other data BuWizz 2 broadcasts (via notification.)

As far as Bleak went, all I've needed is the ability to become a client of a specific server, write to a characteristic, and subscribe to that characteristic for notifications. I'm completely new to this type of thing, but I don't think this brings much at all in the way of bulk. My thought was basically just what was needed for a hub to connect to it and send commands. Motor control is an array of 6 bytes, speed mode is 2. And as its writing to characteristics, you only need to communicate when you want to change the current outputs. The notifications are larger - usually 20 bytes. So I'd probably allow subscribing and unsubscribing, as opposed to assuming the information is wanted. But That is still not much data I suppose - I just prefer to keep a low memory footprint with things unless there's reason not to.

Is there any specific preference for how I'd go about this? Not sure if this will be useful or not, but if it might be, I'd like to do so as well as I can. I'm open to attempting a more generic API, but currently I'm rather limited in the amount of devices I have to test its stability/usability.

**Reply by Mantarri** [reply]
To further expand on my questions [here](https://github.com/orgs/pybricks/discussions/1452#discussioncomment-8580776), which file from the firmware build process is the one that actually gets flashed (i.e. how much storage out of what is available is my build going to use?)

**Reply by dlech** [reply]
The [platform.ld](https://github.com/pybricks/pybricks-micropython/blob/master/lib/pbio/platform/technic_hub/platform.ld) files have the info about how much flash memory is available.

Each time you compile, the makefile prints out the current size. For example:

```
BIN creating firmware base file
199220 bytes
```

was the latest CI build for the Technic hub.

The only hub where where we really have a problem is the Move hub though  (it is currently 100% full).

**Reply by jaguilar** [reply]
I assume the answer to this is yes, but I'm curious if y'all have thoroughly examined the micropython feature set and seen if there is anything that might be turned off to save space. I have no sense of how much or little 200kB is in terms of code. At work, we think about binaries in terms of hundreds of MBs.

Are there any known optimization opportunities y'all just haven't had a chance to get your hands on yet? Or have y'all pretty well combed over the firmware for opportunities and we're already in the best place we can be?

**Reply by dlech** [reply]
> but I'm curious if y'all have thoroughly examined the micropython feature set and seen if there is anything that might be turned off to save space

We started with the minimal MicroPython port (virtually all features turned off) and turned things on as we needed them rather than the other way around. 

> I have no sense of how much or little 200kB is in terms of code.

This is nearly double the usable flash memory on the Move hub and about 94% of the usable flash memory on the Technic hub.

> Are there any known optimization opportunities y'all just haven't had a chance to get your hands on yet? 

https://github.com/pybricks/support/issues/1459

> Or have y'all pretty well combed over the firmware for opportunities and we're already in the best place we can be?

Pretty much. Although we manage to keep eventually find more. It's not easy though.

**Reply by jaguilar** [reply]
> It's not easy though.

I believe it. Possibly this did not come across but I was asking if there is any way someone who is not super familiar with the situation might be able to jump in and help. I'll check out that issue you linked.

**Reply by laurensvalk** [reply]
See also https://stats.pybricks.com/movehub.html for our past endeavors 😄

**Reply by Mantarri** [reply]
Is there a simple example for adding a bluetooth target device to the firmware? I've been crawling through `/pybricks/iodevices/pb_type_iodevices_xbox_controller.c`, it's just a bit hard for me at times figuring out which parts are xbox controller specific, and which are necessary boiler-plate. It doesn't help that I had to disable warning & error squiggles in VS Code, as otherwise it thinks  a lot of headers are missing (is this intentional/due to how the repo's make system is setup, or is there some way I can get [only valid] editor warnings?) Just getting:
```
/opt/gcc-arm-none-eabi-10.3-2021.07/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld: /tmp/firmware.elf.F3yJ8f.ltrans2.ltrans.o:(.rodata.iodevices_globals_table.lto_priv.0+0x24): undefined reference to `pb_type_iodevices_BuWizz2'
```
over and over when compiling isn't giving me much information. 😅

**Reply by laurensvalk** [reply]
Also see the pb_type_lwp3device.c file. 

Undefined reference usually means that the function hasn't been compiled, so the linker can't find it.

I suppose you might not have added your file to the list of files to be compiled yet. See sources.mk for this. 

(in vscode, hit F1 then backspace to search for these files)

**Reply by Mantarri** [reply]
Ah, that was it, thank you, and will also take a look at that file.

**Reply by Mantarri** [reply]
So I've been looking through `pb_type_lwp3device.c`, `bluetooth.h`, and `bluetooth_btstack.c` to figure out how to write to a characteristic and  need some clarification. The only usage I can find is in `pb_type_lwp3device` when using VS Code to search through the files, so I'm not sure if  the usage (`STATIC mp_obj_t lwp3device_write`... line 478) is communicating directly by writing to a characteristic, or if there's some abstraction. Reason being the call to `pbdrv_bluetooth_peripheral_write()` just takes in the `pbio_task_t` member of the `pb_lwp3device_t` struct, and I can't seem to figure out what `pbio_task_t` does/is doing.

TL;DR would be I'm a bit unclear on how to write to a ble device's characteristic (assuming I've connected to it. I've yet to do so as I'm first trying to whittle down my module to the necessities before trying it, to have a less possible places for colliding errors.)

**Reply by Mantarri** [reply]
(Just bumping [this](https://github.com/orgs/pybricks/discussions/1452#discussioncomment-8649257) in case it was missed.)

**Reply by laurensvalk** [reply]
Commands such as `pbdrv_bluetooth_peripheral_write` initiate an operation in the Bluetooth driver (different between most hubs), and the task objects are a way to monitor that operation until it is done, without blocking all the other code that has to run.

The write operation takes a value argument, which contains the handle of the characteristic to write to. In the case of the `lwp3device`, it is set to `pb_lwp3device_char.handle`, which was populated during the discovery process that ran after connecting to the device.
